### PR TITLE
Add creation of new keyboard in child directory

### PR DESCRIPTION
**Pull request by shikamiya**
## Description

~~When you have repo/keyboards/shikamiya directory,~~ you can create new keyboard in keyboards/shikamiya/shikamiya_first with util/new_keyboard.sh script.
(current util/new_keyboard.sh is only supporting direct child of keyboards directory.)

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* None

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by fauxpark** [comment]
Duplicate of #5771

**Comment by shikamiya** [comment]
I'm sorry for duplication.

but I think current script has problem about generating nested keyboard, especially sed and renaming keyboard.{c,h}.
sed error is occurred by command like: `s/%KEYBOARD%/new_keyboard_dir/new_keyboard/g`

when I have keyboards/new_keyboards_dir directory, I got this by current script:

```
Keyboard Name: new_keyboard_dir/new_keyboard
Keyboard Type [avr]: 
Your Name [shikamiya]: 

Copying base template files... done
Copying avr template files... done
Renaming keyboard files...mv: cannot move 'keyboards/new_keyboard_dir/new_keyboard/template.c' to 'keyboards/new_keyboard_dir/new_keyboard/new_keyboard_dir/new_keyboard.c': No such file or directory
mv: cannot move 'keyboards/new_keyboard_dir/new_keyboard/template.h' to 'keyboards/new_keyboard_dir/new_keyboard/new_keyboard_dir/new_keyboard.h': No such file or directory
 done
Replacing %KEYBOARD% with new_keyboard_dir/new_keyboard...sed: -e expression #1, char 31: unknown option to `s'
sed: -e expression #1, char 31: unknown option to `s'
sed: -e expression #1, char 31: unknown option to `s'
sed: -e expression #1, char 31: unknown option to `s'
 done
Replacing %YOUR_NAME% with shikamiya...sed: can't read keyboards/new_keyboard_dir/new_keyboard/new_keyboard_dir/new_keyboard.c: No such file or directory
sed: can't read keyboards/new_keyboard_dir/new_keyboard/new_keyboard_dir/new_keyboard.h: No such file or directory
 done
```

so I just replace keyboard name with keyboard base name.

**Comment by fauxpark** [comment]
This is essentially what the changes I suggested in that issue do, except that it also fixes a bug in the base template copying function, and the PR itself makes sure the directories exist with the `mkdir -p`. So, I think it is a more robust solution.

**Comment by shikamiya** [comment]
I think creating intermediate directory is needed only in first keyboard. usually vendor or user have already their intermediate directory.
and if you need to create new intermediate directory, you can just run mkdir once, and it's not needed again.
and I have an idea with this script: you can commit new empty file `keyboards/test/.keep` and add `/keyboads/test` to gitignore in order to make sure user's testing keyboard won't be included in PRs.

**Comment by fauxpark** [comment]
> I think creating intermediate directory is needed only in first keyboard. usually vendor or user have already their intermediate directory.

`mkdir -p` will handle all cases (entirely new vendor folder, pre-existing vendor folder, creating a new keyboard/revision etc.). It only creates the directories it needs to.

> you can commit new empty file keyboards/test/.keep and add /keyboads/test to gitignore in order to make sure user's testing keyboard won't be included in PRs.

You shouldn't be `git add`ing folders like that in the first place.

**Comment by shikamiya** [comment]
I see. now I added mkdir with message to my branch.

**Comment by shikamiya** [comment]
Do I need to do anything by myself?

**Comment by fauxpark** [comment]
The if statement and the message are not all that useful IMO.

**Comment by shikamiya** [comment]
ok, I removed these message and if statement.

File: util/new_keyboard.sh [hunk 1]
```diff
@@ -32,6 +32,8 @@ set_git_username() {
 
 # Copy the template files to the new keyboard directory.
 copy_templates() {
+    mkdir -p "$(dirname $keyboard_dir)"
```
Comment (fauxpark) [inline util/new_keyboard.sh]:
```suggestion
    mkdir -p "$keyboard_dir"
```

The `dirname` should not be necessary either.

File: util/new_keyboard.sh [hunk 2]
```diff
@@ -32,6 +32,8 @@ set_git_username() {
...
+
     echo -n "Copying base template files..."
     cp -r "quantum/template/base" "${keyboard_dir}"
```
Comment (fauxpark) [inline util/new_keyboard.sh]:
```suggestion
    cp -r "quantum/template/base/." "${keyboard_dir}"
```

A mistake I made when rewriting this script.

File: util/new_keyboard.sh (hunk 1 above)
Comment (shikamiya) [reply util/new_keyboard.sh]: I understood your opinion.
I like copying entire directory, so I kept this `dirname` . it's just my opinion, it doesn't need to keep in qmk repository. I'll fix it.

**Comment by shikamiya** [comment]
thanks for approving, and Do I still have to do?
failing CI? duplicate? still checking?
please tell me whether this PR is stopping by my side!

**Comment by drashna** [comment]
Well, I've retriggered the Travis CI build process, and that should fix the failing. 

Aside from that, it's waiting on somebody with more relevant experience to merge it.

**Comment by drashna** [comment]
Sorry, it looks like there are some merge conflicts here, now.  

Also, we've been trying to move to a more python based solution for this sort of stuff, and it may be worth checking it out.

File: util/new_keyboard.sh [hunk 3]
```diff
@@ -32,17 +32,19 @@ set_git_username() {
...
     echo " done"
 
     echo -n "Copying $keyboard_type template files..."
     cp -r "quantum/template/${keyboard_type}/." "${keyboard_dir}"
     echo " done"
 
     echo -n "Renaming keyboard files..."
-    mv "${keyboard_dir}/keyboard.c" "${keyboard_dir}/${keyboard_name}.c"
-    mv "${keyboard_dir}/keyboard.h" "${keyboard_dir}/${keyboard_name}.h"
+    mv "${keyboard_dir}/template.c" "${keyboard_dir}/${keyboard_base_name}.c"
```
Comment (fauxpark) [inline util/new_keyboard.sh]:
```suggestion
    mv "${keyboard_dir}/keyboard.c" "${keyboard_dir}/${keyboard_base_name}.c"
```

File: util/new_keyboard.sh [hunk 4]
```diff
@@ -32,17 +32,19 @@ set_git_username() {
...
+    mv "${keyboard_dir}/template.h" "${keyboard_dir}/${keyboard_base_name}.h"
```
Comment (fauxpark) [inline util/new_keyboard.sh]:
```suggestion
    mv "${keyboard_dir}/keyboard.h" "${keyboard_dir}/${keyboard_base_name}.h"
```

File: util/new_keyboard.sh (hunk 3 above)
Comment (shikamiya) [reply util/new_keyboard.sh]: fixed :raised_hands:

File: util/new_keyboard.sh (hunk 4 above)
Comment (shikamiya) [reply util/new_keyboard.sh]: fixed :raised_hands:

**Pull request merged**